core/Db.php

```diff
@@ -285,12 +286,55 @@ public static function query($sql, $parameters = array())
         }
     }
 
+    /**
+     * Executes a callback with potential recovery from a "MySQL server has gone away" error.
+     *
+     * If the callback throws a "MySQL server has gone away" exception
+     * it will be called again after a single reconnection attempt.
+     *
+     * @param callable $callback
+     *
+     * @return mixed
+     *
+     * @throws Exception
+     *
+     * @internal
+     */
+    public static function executeWithDatabaseWriterReconnectionAttempt(callable $callback)
+    {
+        try {
+            return $callback();
+        } catch (Exception $ex) {
+            // only attempt reconnection in a reader/writer configuration
+            if (!self::hasReaderConfigured()) {
```

Comment:
Is this covered by tests?

Reply:
There is a [specific test](https://github.com/matomo-org/matomo/pull/21980/files#diff-93a1f0d6b170456ab57a9cf3699a5e1035bbed43951b142a48e387e54cde7513R139) for that case, yes.